### PR TITLE
2D Scatter-Gather

### PR DESCRIPTION
**Pull request by szaman19**
- Adds support for 2-dimensional inputs for the Values into the Scatter and Gather layer
- Updated unit tests to test the new functionality

### CODE REVIEW

File: src/layers/transform/gather.cu [hunk 1]
```diff
@@ -33,87 +33,107 @@ namespace lbann {
...
 
 /** @brief Kernel for scattering a 2D tensor along dim 1
  *
  *  output(j,indices(j,i)) = values(j,i)
  *
- *  Block dimensions: bdimx x bdimy x 1
+ *  Block dimensions: bdimx x bdimy x bdimz
  *
- *  Grid dimensions: (values_dim[1] / bdimx) x (values_dim[0] / bdimy) x 1
+ *  Grid dimensions: (num_columns_input_mat / bdimx) x (num_rows / bdimy) x mb_size /bdimz
```
Comment (timmoon10) [inline src/layers/transform/gather.cu:44]: These grid dimensions aren't used when launching the kernel. It doesn't affect correctness, but will likely hurt performance.

File: src/layers/transform/gather.cu [hunk 2]
```diff
@@ -33,87 +33,107 @@ namespace lbann {
...
 
 /** @brief Kernel for gathering a 2D tensor along dim 1
  *
  *  output(j,i) = values(j,indices(j,i))
  *
- *  Block dimensions: bdimx x bdimy x 1
+ *  Block dimensions: bdimx x bdimy x bdimz
  *
- *  Grid dimensions: (output_dim[1] / bdimx) x (output_dim[0] / bdimy) x 1
+ *  Grid dimensions: (num_columns_output_mat / bdimx) x (num_rows / bdimy) x mb_size /bdimz
```
Comment (timmoon10) [inline src/layers/transform/gather.cu:93]: These grid dimensions aren't used when launching the kernel.

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Overall looks good. Just forgot to update the grid dims in the GPU gather implementation.

File: src/layers/transform/gather.cu (hunk 2 above)
Comment (szaman19) [reply src/layers/transform/gather.cu:93]: Fixed!

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**